run_unittests.py

```diff
@@ -7971,17 +7971,6 @@ def setUp(self):
         src_root = os.path.dirname(__file__)
         self.meson_cross_file = os.path.join(src_root, 'cross', 'ubuntu-armhf.txt')
 
-    def test_cflags_cross_environment_pollution(self):
-        '''
-        Test that the CFLAGS environment variable does not pollute the cross
-        environment. This can't be an ordinary test case because we need to
-        inspect the compiler database.
-        '''
-        testdir = os.path.join(self.common_test_dir, '3 static')
-        self.init(testdir, override_envvars={'CFLAGS': '-DBUILD_ENVIRONMENT_ONLY'})
-        compdb = self.get_compdb()
-        self.assertNotIn('-DBUILD_ENVIRONMENT_ONLY', compdb[0]['command'])
```

Comment:
Waitaniminute. This test actually proves that `CFLAGS` were _not_  put in cross compilation flags earlier. If they had been then this test would have failed. This test has been around since 2017 so that should be established behaviour that people probably depend on.

Reply:
```console
git checkout 0.56.2
cd test\ cases/common/02\ cpp
CXXFLAGS="-m32" ../../../meson.py setup build --cross-file cross.ini
../../../meson.py configure build | grep cpp_args
```

cross.ini
```ini
[binaries]
cpp = 'g++'

[host_machine]
system = 'linux'
endian = 'little'
cpu_family = 'x86'
cpu = 'i686'
```

```
  Compiler options        Current Value                                                                                                                                                                                                                                                                                      >
  ----------------        -------------                                                                                                                                                                                                                                                                                      >
  cpp_args                [-m32]                                                                                                                                                                                                                                                                                             >
  cpp_debugstl            false                                                                                                                                                                                                                                                                                              >
  cpp_eh                  default                                                                                                                                                                                                                                                                                            >
  cpp_link_args           [-m32]                                                                                                                                                                                                                                                                                             >
  cpp_rtti                true                                                                                                                                                                                                                                                                                               >
  cpp_std                 none                                                                                                                                                                                                                                                                                               >
                                                                                                                                                                                                                                                                                                                             >
  build.cpp_args          []                                                                                                                                                                                                                                                                                                 >
  build.cpp_debugstl      false                                                                                                                                                                                                                                                                                              >
  build.cpp_eh            default                                                                                                                                                                                                                                                                                            >
  build.cpp_link_args     []                                                                                                                                                                                                                                                                                                 >
  build.cpp_rtti          true                                                                                                                                                                                                                                                                                               >
  build.cpp_std           none   
```

Reply:
I *think* that test was succeeding before because the cross file has precedence, and was overwriting the CFLAGS values.

Reply:
No, I think that's a bad idea. If we're not going to maintain autotools compatibility (ie, CFLAGS is always for the HOST and CFLAGS_FOR_BUILD is always for the BUILD machine) then we should delete them and not have env var support, full stop. Having them behave differently is terrible UI, and just asking for a ton of bug reports. Otherwise, distros are just going to say "this is annoying", and patch meson to behave the way they expect, and users are going to complain bitterly that "this doesn't work", or "this works fine for me on distro X, don't know distro Y must just be broken".

@eli-schwartz, @Ericson2314, as distro people: opinions?

Reply:
I just realized laying here trying to fall asleep the test i right. There's another big in my implementation... I'll fix it tomorrow.

Reply:
Testing this we find that the current master behaves differently and CFLAGS are not used in cross compilation. Bisection points, as expected, to  2cd0723c42ae2076c7ba2.

Weirdly if I run the test by hand and have things in the cross file's `c_args` the envvar is _still_ put there, even for old versions. I don't understand how the test could pass previously.

Reply:
I'm not completely positive what's being suggested here. Remove $CFLAGS support in cross builds? Remove $CFLAGS support in native builds "because it doesn't work in cross builds"?

Reply:
Removing CFLAGS in cross builds because those should come from the cross file. This is how it was designed and even had tests (the one removed in this MR) to ensure that CFLAGS do not go to cross compilations. However it seems that it was broken somehow and at some point some change made the flags bleed to cross compilations.

This is bad, because for example on Debian the cross file is created on the fly when package building starts. With this behaviour every flag in CFLAGS et al is passed to the compiler twice: once from the cross file and once from the envvar.

Reply:
No, it's not. CFLAGS and machine files are exclusive. If you pass a Machine file with `[built-in options]:c_args` then $CFLAGS will be ignored. *That* was broken, and was why the test was failing. I've put the test back and fixed the implentation such that you only get one or the other.

Reply:
Weird, because I'm fairly sure I tried that with old releases and it did not work. Let me recheck.

Reply:
I could not reproduce it. However upon verification I did discover one discrepancy. If `c_link_args` is not defined in the cross file but `c_args` is, then `CFLAGS` are not added to `c_link_args` whereas it used to be added. This is a good example of why having multiple ways of defining something is not great, because there are so many ways they can be combined that inconsistent behaviour like this appears.

So I'm not really sure what should be done here. Breaking visible behaviour in a point release should be avoided, but on the other hand this is starting to get way too complicated altogether. Suggestions welcome.